rfcs/experimental/flow_graph_wait_try_put_and_wait/README.md

```diff
@@ -0,0 +1,510 @@
+    waiters_type&& waiters() &&;
+
+    bool empty() const;
+
+    void merge(const message_metainfo&);
+};
+```
+
+The current implementation of ``message_metainfo`` class wraps only the list of single message waiters. The class may be extended if necessary to cover additional use-cases.
+
+The metainfo is required to hold a list of message waiters instead of single waiter to cover the ``continue_node`` and ``join_node`` joining use-cases. Consider the example:
+
+```cpp
+using namespace oneapi::tbb;
+
+flow::function_node<int, int> start1(g, ...);
+flow::function_node<int, int> start2(g, ...);
+
+flow::join_node<std::tuple<int, int>> join(g);
+
+flow::function_node<std::tuple<int, int>, int> post_process(g, ...);
+
+flow::make_edge(start, flow::input_port<0>(join));
+flow::make_edge(start, flow::input_port<1>(join));
+flow::make_edge(join, post_process);
+
+std::thread t1([&]() {
+    start1.try_put_and_wait(1);
+});
+
+std::thread t2([&]() {
+    start2.try_put_and_wait(2);
+})
+```
+
+### Combined or separated wait
+
+Current proposal describes only the case where submitting the work into the flow graph and waiting for it are combined in a single use API `node.try_put_and_wait`.
+In theory, it can be useful to also have the ability to split these phases:
+
+```cpp
+
+oneapi::tbb::flow::graph g;
+oneapi::tbb::flow::function_node<int, int> start_node(g, ...);
+
+auto desc = node.try_put(work); // other API name can be selected
+
+// Other work
+// May submit more work into the Flow Graph
+// May create other descriptors that would be waited later
+
+g.wait(desc);
+
+```
+
+In that case it would be needed to extend the node with the new API returning some descriptor that can be used as the argument to the work waiting function `g.wait` that also should be added.
+This descriptor can wrap the metainformation class for the current proposal, but the exact semantics and API should be defined since it makes the metainfo class public in some manner.
+
+### Buffering the metainfo
+
+As described in the [buffering nodes description section](#single-push-buffering-nodes), current proposal requires the nodes that stores the user values as part of buffers to store
+also the metainformation associated with these values in the same buffer for future use and also to hold the additional reference counter to force the `try_put_and_wait` to wait 
+until the rejecting receiver take the item from the buffer and process it. 
+
+Such a behavior may significantly affect the other common use-case for buffering nodes: when they are used to store the result of the computation at the end of the Flow Graph. 
+In such scenarios, the metainformation would be stored in the buffer and never taken from it since the buffering node is not used as part of push-pull protocol and hence there is
+no rejecting successor. The `try_put_and_wait` function associated with such a metainformation will hang forever since the reference counter on the stored metainformation would
```

Comment:
Is `will hang forever` the right description though? I thought that once the corresponding message is taken out of the buffer by some consumer (thread) outside of the graph, the last reference is removed and waiting finishes. So it should be possible to use an asynchronous consumer activity, and the deadlock would only happen if the waiting thread is the consumer, correct?

For such cases, should we consider adding some trait to buffer nodes to ignore the metainformation?